### ISSUE TITLE
How to bake navmesh surface in prefab?

### ISSUE THREAD

**Issue by doraeball22** [Unity-Technologies/NavMeshComponents#86]
How to bake navmesh surface in prefab?

**Comment by keeganstothert** [comment]
I get this error when i try to bake it in the prefab, 

Duplicating NavMeshSurface does not duplicate the referenced navmesh data

**Comment by wrybri** [comment]
Any updates on this? We want to put prebaked navmeshsurfaces into various prefabs for our game

**Comment by turadr** [comment]
In 2018.3 the NavMesh can be baked for a prefab from inside the[ Prefab Mode](https://docs.unity3d.com/Manual/EditingInPrefabMode.html). Is it possible for you to try it and let me know if it works in your case?

**Comment by wrybri** [comment]
Ah-ha! Yes it works. I see that it saves the NavMeshData in the same folder as the prefab, instead of in the scene. I suppose I can do this via editor script too? Ideally I'd like to be able to serialize the NavMeshData for flexibility, to keep it with other serialized data, put it in assetbundles etc etc

**Comment by turadr** [comment]
In the Editor you can save the the NavMeshData asset through scripts. Have a look in `NavMeshAssetManager.cs` if you're looking for an example. https://github.com/Unity-Technologies/NavMeshComponents/blob/d29a5f3e8f163620e658a3d8ae641f0ffe319e93/Assets/NavMeshComponents/Editor/NavMeshAssetManager.cs#L62

The NavMeshData cannot be currently serialized at runtime.

**Comment by Trivvn** [comment]
Would this support instantiating multiple copies of the prefab with the navmesh on it?

For example: Bake the navmesh on a cave prefab, then instantiate the cave in 2 different places with the same navmesh.

**Comment by turadr** [comment]
@Trivvn Yes, multiple instances of the prefab use the same NavMesh asset from the prefab even when they are positioned in different places in the world.

**Comment by dps999** [comment]
Hi, 
I tried to bake prefab in 2018.3.4 and it looks baked in prefab. But when I even drag it to the scene it loses his navmesh and I should click bake again.
Is it a bug or I don't get the point of prefab baking?

**Comment by turadr** [comment]
Could be a bug. The instance in the scene should contain the NavMesh baked in the prefab. Did you save the prefab after baking? Could you capture and share a video where the Scene and Inspector are visible as you are doing these operations? Thanks.

**Comment by dps999** [comment]
Sorry, I have updated the NavMeshComponents and it works now.

**Comment by S786786786** [comment]
How to bake navmesh surface in prefab in unity 2019 or 2020?

**Comment by fkkcloud** [comment]
@turadr Is it possible to bake the navmesh in PrefabMode then use it in a scene? Same Instance, different multiple positions in 2019.4.17f1 LTS?
It seems like the navmesh does not follow its parent transform position when I drag the prefab around.

**Comment by turadr** [comment]
@fkkcloud Yes, you can bake a NavMesh in prefab mode. You save the prefab and then all the instances of the prefab carry their own instance of the NavMesh following the transform of the game object. Do NavMeshAgents not run correctly on these instances when you start the game? If it does not work for you please [report a bug](https://unity3d.com/unity/qa/bug-reporting) so that our Quality Assurance department can assist you. 

If you checkout/download the [surface-prefab-example](https://github.com/Unity-Technologies/NavMeshComponents/tree/surface-prefab-example) branch of this repository then you can run the [7_dungeon.unity](https://github.com/Unity-Technologies/NavMeshComponents/blob/surface-prefab-example/Assets/Examples/Scenes/7_dungeon.unity) scene where the maze is built from prefab instances.

**Comment by alexknutson** [comment]
Just wanted to jump in here and thank you @turadr for linking to the surface-prefab-example project. I wasn't aware of it, and I've been under the impression that pre-baking nav mesh surface data on a prefab was impossible, so I've been doing all the heavy lifting at runtime. 🍻 cheers!

**Comment by fkkcloud** [comment]
@turadr thank you. A quick question. Is it possible to not make the Agent to jump in the NavmeshLink? Can they just walk through it?

**Comment by turadr** [comment]
You're welcome. :)
In the Dungeon example the agent object has a script component called `Agent Link Mover` that controls the movement over the link. You can select the `Normal Speed` method in order to have it "walk" over the links.

Generally, if you tick the agent's `Auto Traverse OffMeshLinks` option then the agent walks but with double of its maximum speed and the Link Mover script does nothing.

![image](https://user-images.githubusercontent.com/9038468/111967840-cfeefa80-8af8-11eb-9121-1f71adfd7646.png)

**Comment by trsh** [comment]
How to bake navmesh surface in prefab in unity 2021?

**Comment by turadr** [comment]
We are aware of the issue that the NavMesh is not displayed when editing a prefab _in isolation_ and we have been trying to address it but I don't have any ETA for a fix yet. The NavMesh will be generated, but not displayed. An alternative for displaying the NavMesh is to edit the prefab _in context_. The two modes are described here https://docs.unity3d.com/Manual/EditingInPrefabMode.html .

Regarding the two errors, they are not produced by the Navigation system but by the asset import pipeline (possibly in conjunction with UI Toolkit). The issue is tracked internally by another team and they don't have a fix for it yet. Maybe in the future they'll provide an update in the forum thread here https://forum.unity.com/threads/error-on-fresh-project-w-ui-toolkit-2021-1-10-2062.1072106/ .
